### ISSUE TITLE
Serialization of a sol::state ?

### ISSUE THREAD

**Issue by FHedin** [ThePhD/sol2#606]
Hi all,

I would like to know if someone already attempted to serialize a sol::state successfully : I am developing an HPC application with a limited runtime of 24 hours, and I need to serialize the state of my program for a later restart.
In my C++ app I therefore have a `sol::state lua;` containing many functions and variables; do you know an efficient way to serialize this sol::state ? or at least a part of it ?

The following link http://lua-users.org/wiki/TableSerialization discusses tables serialization in Lua, I am ready to try several of them, but in case someone already knows a working solution, I would appreciate saving some time ;-)

Thanks a lot

**Comment by RussellHaley** [comment]
I have had good success over the last few years with serpent. It's been working fine with sol but has limitations for advanced serialization (i.e functions, userdata). Penlight is a good general purpose library that has serialization among other great tools.

**Comment by ThePhD** [comment]
I'm gunshy of making a "Deep Copy" or "Deep Serialization" recommendation because at the end of the day -- due to functions and userdata -- only _you_ know how to properly serialize your stuff. Regular tables, integers, strings, etc. are all fine, but Userdata have custom layout and everything else.

If you can find some way of providing a standard hook for serialize/deserialize on your userdata, for example, then you can write a recursive function to also write those out. I have **no** idea what you'd do for a function.

Good luck!

**Comment by FHedin** [comment]
Hi all,

thanks for those suggestions, in the end I have found a different approach which is not related at all to any Lua library, but it might be of interest for some of you so I will still share it within this thread.

I am using the DMTCP tool (https://github.com/dmtcp/dmtcp), performing a full check-pointing of my C++ application: it will dump to a file the full state and memory space of the application at a given time point, and allow to re-execute later the program and continue its execution. The `sol::state` appears to be fully preserved, even the `sol::function`'s I have used are still valid and successfully called from the C++ side !

And according to the documentation it is even possible to migrate the files to a different machine with a different software environment (as long as the cpu arch is compatible of course) because dependencies such as *.so are also included in the check-pointing dump process. However I haven't tried this feature yet.

I hope this can be useful to someone else.
